packages/cozy-harvest-lib/src/services/budget-insight.js

```diff
@@ -244,35 +244,49 @@ export const resumeBIConnection = flow => {
 }
 
 /**
- * Checks for 2FA or decoupled requests
+ * @typedef biConnection
+ * @property {number} id
+ * @property {number} id_user
+ * @property {number} id_connector
+ * @property {string|null} state  - ( wrongpass | additionalInformationNeeded | websiteUnavailable | actionNeeded | SCARequired | decoupled | passwordExpired | webauthRequired | rateLimiting | bug). Null indicates a success
+ * @property {string|null} last_update - Date string: Last successful update.
+ * @property {string|null} created - Date string: Creation date
+ * @property {boolean} active - Whether this connection is active and will be automatically synced.
+ * @property {string|null} last_push - Date string: Last successfull push
+ * @property {string|null} next_try - Date string: Date of next synchronization.
+ */
+
+/**
+ * Checks for any number 2FA and/or decoupled requests
  *
  * - Resolves when 2FA has been resolved
  * - Triggers LOGIN_SUCCESS_EVENT
+ *
+ * @param  {Object} options
+ * @param  {biConnection} options.biConnection
+ * @param  {ConnectionFlow} options.flow
  */
-export const finishConnection = async ({ biConnection, flow }) => {
-  if (biConnection.error) {
-    if (biConnection.error === DECOUPLED_ERROR) {
+export const finishConnection = async ({ flow }) => {
+  while (flow.getData().biConnection.error) {
+    const connectionError = flow.getData().biConnection.error
```

Comment:
Nit: connectionError could be declared with a let outside the loop not to have twice the `flow.getData().biConnection.error` assignment.

```
let connectionError
while (connectionError = flow.getData().biConnection.error) {
  ...

  // later on
  connectionError = flow.getData().biConnection.error
  throw convertBIErrortoKonnectorJobError(connectionError)
}
```